File: js/src/blob.ts
/* Blob — an owning byte buffer wrapping a msgpack-encoded value. */

import { Buf } from "./format.ts";
import * as D from "./decode.ts";
import * as E from "./encode.ts";
import * as J from "./json.ts";
import * as M from "./mutate.ts";
import { Type, Value } from "./value.ts";

export class Blob {
  _data: Uint8Array;

  constructor(data?: Uint8Array | number[] | null) {
    if (data === null || data === undefined) this._data = new Uint8Array(0);
    else if (data instanceof Uint8Array) this._data = data;
    else this._data = Uint8Array.from(data);
  }

  // ── raw access ────────────────────────────────────────────────────
  data(): Uint8Array {
    return this._data;
  }
  size(): number {
    return this._data.length;
  }
  empty(): boolean {
    return this._data.length === 0;
  }
  hex(): string {
    let s = "";
    for (let i = 0; i < this._data.length; i++) s += this._data[i].toString(16).padStart(2, "0");
    return s;
  }
  equals(other: Blob): boolean {
    if (this._data.length !== other._data.length) return false;
    for (let i = 0; i < this._data.length; i++) if (this._data[i] !== other._data[i]) return false;
    return true;
  }

  // ── validation ────────────────────────────────────────────────────
  valid(): boolean {
    return D.isValid(this._data, this._data.length);
  }
  errorPosition(): number {
    return D.errorPosition(this._data, this._data.length);
  }

  // ── type inspection ───────────────────────────────────────────────
  type(path?: string): Type {
    const n = this._data.length;
    if (path === undefined) return n === 0 ? Type.Nil : D.getType(this._data, n, 0);
    const r = D.lookup(this._data, n, 0, path);
    if (r.rc !== D.RC_OK) return Type.Nil;
    return D.getType(this._data, n, r.iStart);
  }
  typeStr(path?: string): string {
    return this.type(path);
  }

  // ── extraction ────────────────────────────────────────────────────
  extract(path: string): Value {
    const n = this._data.length;
    const r = D.lookup(this._data, n, 0, path);
    if (r.rc !== D.RC_OK) return Value.nil();
    return D.decodeElement(this._data, n, r.iStart, r.iEnd);
  }

  arrayLength(path?: string): number {
    const n = this._data.length;
    if (path === undefined) return n === 0 ? -1 : D.getContainerCount(this._data, n, 0);
    const r = D.lookup(this._data, n, 0, path);
    if (r.rc !== D.RC_OK) return -1;
    return D.getContainerCount(this._data, n, r.iStart);
  }

  // ── mutation (copy-on-write) ──────────────────────────────────────
  _apply(path: string, value: Value, mode: number): Blob {
    const nb = new Buf();
    E.encodeValue(nb, value);
    const r = M.applyEdit(this._data, this._data.length, path, nb.toBytes(), mode);
    return r.rc === M.RC_OK ? new Blob(r.out) : this;
  }

  set(path: string, value: Value | Blob): Blob {
    if (value instanceof Blob) {
      const r = M.applyEdit(this._data, this._data.length, path, value._data, M.EDIT_SET);
      return r.rc === M.RC_OK ? new Blob(r.out) : this;
    }
    return this._apply(path, value, M.EDIT_SET);
  }
  insert(path: string, value: Value): Blob {
    return this._apply(path, value, M.EDIT_INSERT);
  }
  replace(path: string, value: Value): Blob {
    return this._apply(path, value, M.EDIT_REPLACE);
  }
  arrayInsert(path: string, value: Value): Blob {
    return this._apply(path, value, M.EDIT_ARRAY_INS);
  }
  remove(path: string): Blob {
    const r = M.applyEdit(this._data, this._data.length, path, new Uint8Array(0), M.EDIT_REMOVE);
    return r.rc === M.RC_OK ? new Blob(r.out) : this;
  }
  patch(mergePatch: Blob): Blob {
    const r = M.mergePatch(this._data, this._data.length, 0, mergePatch._data, mergePatch._data.length, 0, 0);
    return r.rc === M.RC_OK ? new Blob(r.out) : this;
  }

  // ── JSON conversion ───────────────────────────────────────────────
  toJson(): string {
    if (this._data.length === 0) return "null";
    return J.toJson(this._data, this._data.length, false, 0);
  }
  toJsonPretty(indent: number = 2): string {
    if (this._data.length === 0) return "null";
    indent = Math.max(0, Math.min(8, indent));
    return J.toJson(this._data, this._data.length, true, indent);
  }

  static fromJson(json: string | Uint8Array | null | undefined): Blob {
    return new Blob(J.fromJson(json));
  }
}

File: rust/src/blob.rs
//! `Blob` — an owning byte buffer wrapping a msgpack-encoded value.

use crate::decode as d;
use crate::encode as e;
use crate::json as j;
use crate::mutate as m;
use crate::value::{Type, Value};

/// A MessagePack BLOB supporting read, mutation (copy-on-write) and JSON.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct Blob {
    data: Vec<u8>,
}

impl Blob {
    /// Construct from raw bytes (copies).
    pub fn new(data: &[u8]) -> Blob {
        Blob {
            data: data.to_vec(),
        }
    }
    /// Construct by taking ownership of a byte vector.
    pub fn from_vec(data: Vec<u8>) -> Blob {
        Blob { data }
    }

    // ── raw access ────────────────────────────────────────────────────
    pub fn data(&self) -> &[u8] {
        &self.data
    }
    pub fn size(&self) -> usize {
        self.data.len()
    }
    pub fn is_empty(&self) -> bool {
        self.data.is_empty()
    }
    pub fn hex(&self) -> String {
        let mut s = String::with_capacity(self.data.len() * 2);
        for &b in &self.data {
            s.push_str(&format!("{:02x}", b));
        }
        s
    }

    // ── validation ────────────────────────────────────────────────────
    pub fn valid(&self) -> bool {
        d::is_valid(&self.data, self.data.len())
    }
    pub fn error_position(&self) -> usize {
        d::error_position(&self.data, self.data.len())
    }

    // ── type inspection ───────────────────────────────────────────────
    /// Type of the root element.
    pub fn root_type(&self) -> Type {
        if self.data.is_empty() {
            Type::Nil
        } else {
            d::get_type(&self.data, self.data.len(), 0)
        }
    }
    /// Type of the element at `path`.
    pub fn type_at(&self, path: &str) -> Type {
        let n = self.data.len();
        let (rc, istart, _) = d::lookup(&self.data, n, 0, path);
        if rc != d::RC_OK {
            Type::Nil
        } else {
            d::get_type(&self.data, n, istart)
        }
    }
    pub fn type_str(&self) -> &'static str {
        crate::value::type_str(self.root_type())
    }
    pub fn type_str_at(&self, path: &str) -> &'static str {
        crate::value::type_str(self.type_at(path))
    }

    // ── extraction ────────────────────────────────────────────────────
    pub fn extract(&self, path: &str) -> Value {
        let n = self.data.len();
        let (rc, istart, iend) = d::lookup(&self.data, n, 0, path);
        if rc != d::RC_OK {
            Value::nil()
        } else {
            d::decode_element(&self.data, n, istart, iend)
        }
    }

    /// Element count of the root container, or `-1`.
    pub fn array_length(&self) -> i64 {
        if self.data.is_empty() {
            -1
        } else {
            d::get_container_count(&self.data, self.data.len(), 0)
        }
    }
    /// Element count of the container at `path`, or `-1`.
    pub fn array_length_at(&self, path: &str) -> i64 {
        let n = self.data.len();
        let (rc, istart, _) = d::lookup(&self.data, n, 0, path);
        if rc != d::RC_OK {
            -1
        } else {
            d::get_container_count(&self.data, n, istart)
        }
    }

    // ── mutation (copy-on-write) ──────────────────────────────────────
    fn apply(&self, path: &str, value: &Value, mode: i32) -> Blob {
        let mut nb = Vec::new();
        e::encode_value(&mut nb, value);
        let (rc, out) = m::apply_edit(&self.data, self.data.len(), path, &nb, mode);
        if rc == m::RC_OK {
            Blob::from_vec(out)
        } else {
            self.clone()
        }
    }

    pub fn set(&self, path: &str, value: &Value) -> Blob {
        self.apply(path, value, m::EDIT_SET)
    }
    /// Set `path` to an existing sub-blob (embedded verbatim).
    pub fn set_blob(&self, path: &str, sub: &Blob) -> Blob {
        let (rc, out) = m::apply_edit(&self.data, self.data.len(), path, &sub.data, m::EDIT_SET);
        if rc == m::RC_OK {
            Blob::from_vec(out)
        } else {
            self.clone()
        }
    }
    pub fn insert(&self, path: &str, value: &Value) -> Blob {
        self.apply(path, value, m::EDIT_INSERT)
    }
    pub fn replace(&self, path: &str, value: &Value) -> Blob {
        self.apply(path, value, m::EDIT_REPLACE)
    }
    pub fn array_insert(&self, path: &str, value: &Value) -> Blob {
        self.apply(path, value, m::EDIT_ARRAY_INS)
    }
    pub fn remove(&self, path: &str) -> Blob {
        let (rc, out) = m::apply_edit(&self.data, self.data.len(), path, &[], m::EDIT_REMOVE);
        if rc == m::RC_OK {
            Blob::from_vec(out)
        } else {
            self.clone()
        }
    }
    pub fn patch(&self, merge_patch: &Blob) -> Blob {
        let (rc, out) = m::merge_patch(
            &self.data,
            self.data.len(),
            0,
            &merge_patch.data,
            merge_patch.data.len(),
            0,
        );
        if rc == m::RC_OK {
            Blob::from_vec(out)
        } else {
            self.clone()
        }
    }

    // ── JSON conversion ───────────────────────────────────────────────
    /// Byte-exact JSON serialisation (may contain non-UTF-8 string bytes,
    /// identical to the C++ library output).
    pub fn to_json_bytes(&self) -> Vec<u8> {
        if self.data.is_empty() {
            return b"null".to_vec();
        }
        j::to_json_bytes(&self.data, self.data.len(), false, 0)
    }
    /// JSON serialisation as a `String` (exact for valid UTF-8 input).
    pub fn to_json(&self) -> String {
        bytes_to_string(self.to_json_bytes())
    }
    pub fn to_json_pretty_bytes(&self, indent: i32) -> Vec<u8> {
        if self.data.is_empty() {
            return b"null".to_vec();
        }
        let indent = indent.clamp(0, 8);
        j::to_json_bytes(&self.data, self.data.len(), true, indent)
    }
    pub fn to_json_pretty(&self, indent: i32) -> String {
        bytes_to_string(self.to_json_pretty_bytes(indent))
    }

    /// Parse JSON bytes into a msgpack blob.
    pub fn from_json_bytes(json: &[u8]) -> Blob {
        Blob::from_vec(j::from_json(json))
    }
    /// Parse a JSON string into a msgpack blob.
    pub fn from_json(json: &str) -> Blob {
        Blob::from_vec(j::from_json(json.as_bytes()))
    }
}

fn bytes_to_string(b: Vec<u8>) -> String {
    match String::from_utf8(b) {
        Ok(s) => s,
        Err(e) => String::from_utf8_lossy(e.as_bytes()).into_owned(),
    }
}
